TickSpec/ServiceProvider.fs

```diff
@@ -1,10 +1,50 @@
 ﻿namespace TickSpec
 
 open System
-open System.Collections.Generic
 open System.Diagnostics
 open System.Reflection
 
+/// Instances storage
+[<Sealed>]
+type InstanceStore () =
+    let mutable instances : list<Type*obj> = []
+
+    /// Prepends key/value to list if not present; existing key with different value is removed
+    let distinctAdd key value inList =
+        let rec distinctAddRec key value revHead inTail =
+            match inTail with
+            | [] -> (key, value) :: inList
+            | (t, instance:obj) :: rest when key = t ->
+                if value <> instance then
+                    match instance with
+                    | :? IDisposable as d -> d.Dispose()
+                    | _ -> ()
+                    (key, value) :: (List.rev revHead) @ rest
+                else
+                    inList
+            | head :: rest ->
+                distinctAddRec key value (head :: revHead) rest
+        distinctAddRec key value [] inList
+
+    /// Stores element (skips storing existing, old element removed and new prepended)
+    member this.Store key value =
+        instances <- instances |> distinctAdd key value
```

Comment:
should only store if `:? IDisposable`

Reply:
The store serves also as the lookup. So, it has to store everything.